### ISSUE TITLE
RFC: Complex operators (Sort, Cholesky, FFT, ScatterAdd, etc.)

### ISSUE THREAD

**Issue by ekzhang** [ekzhang/jax-js#20]
Right now the `Backend` interface is built around an `AluExp`-centric interface: given an expression, compute that expression across `size` elements, optionally with a reduction. This is encoded in the `Kernel` class, and you can dispatch `Kernel` to a backend.

```ts
/** A device backend. */
export interface Backend {
  // ... <abridged>

  /** Prepare an expression to be executed later. */
  prepare(kernel: Kernel): Promise<Executable>;

  /** Prepare an expression to be executed later, blocking variant. */
  prepareSync(kernel: Kernel): Executable;

  /**
   * Run a backend operation that was previously prepared.
   *
   * The operation may not run immediately, but operations are guaranteed to run
   * in the dispatch order. Also, `read()` will wait for all pending operations
   * on that slot to finish.
   */
  dispatch(exe: Executable, inputs: Slot[], outputs: Slot[]): void;
}
```

This can support like [80% of NumPy operations](https://github.com/ekzhang/jax-js/blob/main/FEATURES.md) well, including matrix multiplication, elementwise ops, and basically everything that most neural networks need to run (except MoE). See [DETR ResNet-50](https://jax-js.com/detr-resnet-50) or [MobileCLIP2](https://jax-js.com/mobileclip) demos for instance. The MFU here is quite good.

In the process of implementing all of the APIs though, it seems some remaining 20% of operations are difficult to support. Here's a representative list.

| Operation | Difficulty | Hack to support it |
| --- | --- | --- |
| Sort | Partition/sort/topk etc. are algorithms like bitonic mergesort, have data dependencies | Bitonic sort with careful merge ops |
| Cholesky | LAPACK osp + triangular solve need iteration | Introduce the `jax.lax.scan` op and dispatch N kernels, not great perf |
| FFT | Iteration of log(n) steps, also many kernel dispatches | Improve JIT to fuse more and reduce compile overhead |
| ScatterAdd | `atomicAdd()` special case | Add an extra feature flag to `Kernel` just for ScatterAdd op, to compute index as an expression |
| SDPA/GQA | easy to implement now, hard to implement "FlashAttention" fast, fused softmax | add another Primitive just for SDPA with a fast lowering path |

The ScatterAdd hack seems fine, but it's a significant shared code complexity for just one kernel. And the other ops do not have suitable hacks that have good performance.

As a reminder, jax-js is an individual project. It's already quite large for an individual project (~25K lines of code), and the complexity is very high in supporting the entire API of a library like NumPy or JAX, while also leaving time/room to help optimize and keep things performant.

We do have the advantage of following in the footsteps of people who have "figured it out" in the past though, as in Tinygrad or JAX. But this is an instance where we don't have a great example to follow — Tinygrad is able to do [sort](https://docs.tinygrad.org/tensor/ops/#tinygrad.Tensor.sort) with a lot of careful shape index plumbing, but the WebGPU dispatch overhead from JS is much slower than that of CUDA/Metal in Python, and it's also a lot of work for them to achieve such good optimization of their IR (despite the name, Tinygrad is very sophisticated). jax-js's IR at present is much simpler.

I'm thinking, this may be the right time to introduce _custom ops_ along with an API to write per-device kernels.

### Custom ops

Basically, you'll define a custom operation with its own dispatch rule and self-contained jvp/vjp rules.

The scope here will be to define a `Primitive.CustomOp` operation with an _ID_ in the params. For example:

```ts
bind(Primitive.CustomOp, [arg1], { name: "linalg.cholesky" });
```

Then, the library (or a user, eventually!) would be able to specify custom ops, perhaps in a `custom-ops` folder:

```
src/
  custom-ops/
    attention.ts
    cholesky.ts
    fft.ts
    scatter-add.ts
    sort.ts
```

We can then implement each of these primitives in its own file, along with custom kernel code for each backend.

- **CPU:** Just write a function that takes `[Tensor] -> [Tensor]`. Here, `Tensor` = JS `TypedArray` + our `ShapedArray`
- **Wasm:** Will need to provide a `CodeGenerator` object for emitting bytecode.
- **WebGPU:** Write a GPU shader, with some `ShapedArray` parameters denoting the inputs and outputs.

I think this is an acceptable compromise — it allows us to define individual operations lowered to certain devices, skipping the `AluExp` part, so we don't need to make our backend compiler IR significantly more complex either.

The downside is that we will need to start exposing parts of the backend codegen as libraries.

### More details

There's quite a bit of design to flesh out here (e.g., closure under VJP/JVP), I think we should review some docs and what others have done in the past carefully.

- [JAX's custom_vjp / custom_jvp doc](https://docs.jax.dev/en/latest/jep/2026-custom-derivatives.html)
- [Custom extensions in MLX](https://ml-explore.github.io/mlx/build/html/dev/extensions.html)

But what we have already is enough to get started coding & seeing what happens.

**Comment by ekzhang** [comment]
This is great work @luisdiaz1997, I think it's what we would be looking for as a custom op. I also see you figured out how to get it running on WebGPU which is really impressive.

I want to take your CPU/WebGPU impls and see how we can move them into a "custom op" framework rather than all living in the `array.ts` file. Currently there's no backend-specific code in `array.ts` as well, so I feel now wouldn't be the time to start.

<img width="648" height="502" alt="Image" src="https://github.com/user-attachments/assets/b5baa0c0-b75a-45d5-ad37-eb3224d7f6ab" />

I also feel some of the code in this PR is a bit duplicated compared to the optimized impls in `numpy.ts`. Perhaps we can figure out how to get the dependencies working for custom JVP rules so that they can use NumPy functions as well — or we move `lax.dot()` up a level so others can depend on it.

This is an excellent starting point
